File: python/afgame_test.py
__author__ = 'pradermecker'

import unittest
from afgame import *

def new_last_frame_game():
    game = Game()
    game.frames = [Frame(), Frame(), Frame(), Frame(), LastFrame()]
    return game

class TestGames(unittest.TestCase):
    def testRefuseExtraIllegalShots(self):
        self.assertEqual(new_last_frame_game().shot(1, 2, 3, 4), 6)
        self.assertEqual(new_last_frame_game().shot(ALL_PINS, 1, 2, 3, 4), 21)
        self.assertEqual(new_last_frame_game().shot(ALL_PINS, ALL_PINS, 1, 2, 3), 33)

    def testGames(self):
        self.assertEqual(Game().shot(ALL_PINS, 8, 1, 2, 1, 2, (ALL_PINS - 2 - 1), 6, 4, 1, ALL_PINS, 8, 2, 3), 101)
        self.assertEqual(Game().shot(8, 1, 1, 8, (ALL_PINS - 8), 1, 2, 1, ALL_PINS, 1, 2, 1), 55)
        self.assertEqual(Game().shot(ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS, 1, 2, 2), 164)

    def testPerfectGame(self):
        self.assertEqual(Game().shot(ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS),
                         300)

    def testLastFrame(self):
        # With one spare
        self.assertEqual(new_last_frame_game().shot(8, (ALL_PINS - 8), 1, 2), 18)
        # With a strike and a spare
        self.assertEqual(new_last_frame_game().shot(ALL_PINS, 10, ALL_PINS - 10, 2), 32)
        # With a strike and a spare at last position
        self.assertEqual(Game().shot(ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS, ALL_PINS, 10, 2, 3), 247)

if __name__ == '__main__':
    unittest.main()

  

File: go/afgame_test.go
package afgame

import (
	"testing"
)

type fix struct {
	label string
	in    []int
	out   int
}

var fixes = []fix{
	fix{"Simple game", []int{NumPin, 8, 1, 2, 1, 2, 12, 6, 4, 1, NumPin, 8, 2, 3}, 101},
	fix{"Spare game", []int{8, 1, 1, 8, (NumPin - 8), 1, 2, 1, NumPin, 1, 2, 1}, 55},
	fix{"Perfect game", []int{NumPin, NumPin, NumPin, NumPin, NumPin, NumPin, NumPin, NumPin}, 300},
	fix{"Refuse extra shot", playLastFrame(1, 2, 3, 4), 6},
	fix{"Refuse extra shot with 1 strike", playLastFrame(NumPin, 1, 2, 3, 4), 21},
	fix{"Refuse extra shot with 2 strike", playLastFrame(NumPin, NumPin, 1, 2, 3), 33},
	fix{"Last frame with a strike and a spare", playLastFrame(NumPin, 10, (NumPin - 10), 2), 32},
	fix{"Last frame with a strike and a spare at last position", playLastFrame(NumPin, 10, 2, 3), 30},
}

func playLastFrame(shots ...int) []int {
	return append([]int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}, shots...)
}

func TestAll(t *testing.T) {
	for _, fix := range fixes {
		game := New()
		v := game.Shots(fix.in...)
		if v != fix.out {
			t.Errorf("%q %d = %d, want %d.", fix.label, fix.in, v, fix.out)
			t.Log(game.frames)
		}
	}
}

